### PR TITLE
Remove SILGenSILBuilder

### PR DESCRIPTION
**Pull request by slavapestov**
The ClangImporter synthesizes a large number of conformances for imported types, and we want to emit them lazily where possible, to speed up compile time and reduce binary size.

The existing lazy conformance forcing in SILGenSILBuilder was very incomplete, and was added to fill in the gaps for the complementary `SourceFile::getUsedConformances()` mechanism.

This PR replaces SILGenSILBuilder with a post-processing pass that walks all the SILInstructions in a function. It handles more instructions.

This is all in preparation for removing `SourceFile::getUsedConformances()` and some other related broken crap.

### CODE REVIEW

**Comment by slavapestov** [comment]
@swift-ci Please test

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Comment by slavapestov** [comment]
@swift-ci Please test compiler performance

**Comment by slavapestov** [comment]
@swift-ci Please benchmark

**Comment by slavapestov** [comment]
@swift-ci Please test

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Comment by slavapestov** [comment]
@swift-ci Please test compiler performance

**Comment by slavapestov** [comment]
@swift-ci Please benchmark

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Comment by slavapestov** [comment]
@swift-ci Please test

**Comment by slavapestov** [comment]
@swift-ci Please smoke test macOS

File: lib/SILGen/SILGenApply.cpp [hunk 1]
```diff
@@ -1781,21 +1781,19 @@ static void emitRawApply(SILGenFunction &SGF,
   }
 
   auto resultType = substFnConv.getSILResultType();
-  auto calleeType = SILType::getPrimitiveObjectType(substFnType);
 
   // If the function is a coroutine, we need to use 'begin_apply'.
   if (substFnType->isCoroutine()) {
     assert(!substFnType->hasErrorResult());
-    auto apply = SGF.B.createBeginApply(loc, fnValue, subs, argValues);
+    auto apply = SGF.B.createBeginApply(loc, fnValue, subs, argValues, false);
```
Comment (DougGregor) [inline lib/SILGen/SILGenApply.cpp]: I'd prefer to have a `/*isNonThrowing=*/` on these `false` arguments.

File: lib/SILGen/SILGenBridging.cpp [hunk 2]
```diff
@@ -946,6 +937,7 @@ SILGenFunction::emitBlockToFunc(SILLocation loc,
     auto loc = RegularLocation::getAutoGeneratedLocation();
     buildBlockToFuncThunkBody(thunkSGF, loc, blockType, funcType,
                               loweredBlockTy, loweredFuncTyWithoutNoEscape);
+    SGM.emitLazyConformancesForFunction(thunk);
```
Comment (DougGregor) [inline lib/SILGen/SILGenBridging.cpp]: There are a lot of these `emitLazyConformancesForFunction` calls scattered around, and it seems easy to forget it when we add the next function that creates a new SIL function body. Could we put a NDEBUG-only check in the `SILGenFunction` constructor to ensure that `emitLazyConformancesForFunction` was called on that function? Or go really wild and call `emitLazyConformancesForFunction` from the destructor?

**Review by DougGregor — APPROVED** [review]

**Comment by slavapestov** [comment]
@DougGregor Thanks for the review! Since I'm running a 'please test compiler performance' for the last 9 hours (ugh!) I don't want to start it again so I'll address your feedback in the next PR.

**Comment by slavapestov** [comment]
@gottesmm In the long run I want IRGen to be able to request SILGen to emit these witness tables on demand. Even wilder.

**Comment by slavapestov** [comment]
# Summary for master full

**Unexpected test results, excluded stats for ProcedureKit, Tagged, Wordy, Deferred**

**Regressions found (see below)**

- [Debug-batch](#debug-batch-39e557b914ed98a1)
    - [debug-batch brief](#debug-batch-brief-39e557b914ed98a1)
    - [debug-batch detailed](#debug-batch-detailed-39e557b914ed98a1)
- [Release](#release-39e557b914ed98a1)
    - [release brief](#release-brief-39e557b914ed98a1)
    - [release detailed](#release-detailed-39e557b914ed98a1)



<a name="debug-batch-39e557b914ed98a1">

# Debug-batch

<a name="debug-batch-brief-39e557b914ed98a1">

## debug-batch brief
<details>
<summary>Regressed (0)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
</details>
<details>
<summary>Improved (0)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
</details>
<details>
<summary>Unchanged (delta < 1.0% or delta < 100.0ms) (3)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
Frontend.NumInstructionsExecuted | 29,797,283,768,170 | 29,714,923,670,196 | -82,360,097,974 | -0.28%
LLVM.NumLLVMBytesOutput | 1,114,729,562 | 1,114,726,372 | -3,190 | -0.0%
time.swift-driver.wall | 2749.9s | 2760.5s | 10.6s | 0.39%
</details>

<a name="debug-batch-detailed-39e557b914ed98a1">

## debug-batch detailed
<details>
<summary>Regressed (2)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
Driver.NumDriverPipePolls | 197,910 | 201,100 | 3,190 | 1.61% :no_entry:
Driver.NumDriverPipeReads | 223,036 | 225,583 | 2,547 | 1.14% :no_entry:
</details>
<details>
<summary>Improved (0)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
</details>
<details>
<summary>Unchanged (delta < 1.0% or delta < 100.0ms) (104)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
AST.NumASTBytesAllocated | 72,574,566,324 | 72,562,789,625 | -11,776,699 | -0.02%
AST.NumDecls | 92,845 | 92,845 | 0 | 0.0%
AST.NumDependencies | 212,126 | 212,126 | 0 | 0.0%
AST.NumImportedExternalDefinitions | 1,199,385 | 1,199,385 | 0 | 0.0%
AST.NumInfixOperators | 34,221 | 34,221 | 0 | 0.0%
AST.NumLinkLibraries | 0 | 0 | 0 | 0.0%
AST.NumLoadedModules | 253,056 | 253,056 | 0 | 0.0%
AST.NumLocalTypeDecls | 123 | 123 | 0 | 0.0%
AST.NumObjCMethods | 15,429 | 15,429 | 0 | 0.0%
AST.NumPostfixOperators | 18 | 18 | 0 | 0.0%
AST.NumPrecedenceGroups | 16,906 | 16,906 | 0 | 0.0%
AST.NumPrefixOperators | 85 | 85 | 0 | 0.0%
AST.NumReferencedDynamicNames | 122 | 122 | 0 | 0.0%
AST.NumReferencedMemberNames | 4,049,560 | 4,049,560 | 0 | 0.0%
AST.NumReferencedTopLevelNames | 312,993 | 312,993 | 0 | 0.0%
AST.NumSourceBuffers | 387,344 | 387,344 | 0 | 0.0%
AST.NumSourceLines | 2,974,061 | 2,974,061 | 0 | 0.0%
AST.NumSourceLinesPerSecond | 2,374,756 | 2,369,786 | -4,970 | -0.21%
AST.NumTotalClangImportedEntities | 4,582,034 | 4,568,103 | -13,931 | -0.3%
AST.NumUsedConformances | 260,454 | 260,454 | 0 | 0.0%
Driver.ChildrenMaxRSS | 110,466,719,744 | 110,645,968,896 | 179,249,152 | 0.16%
Driver.DriverDepCascadingDynamic | 0 | 0 | 0 | 0.0%
Driver.DriverDepCascadingExternal | 0 | 0 | 0 | 0.0%
Driver.DriverDepCascadingMember | 0 | 0 | 0 | 0.0%
Driver.DriverDepCascadingNominal | 0 | 0 | 0 | 0.0%
Driver.DriverDepCascadingTopLevel | 0 | 0 | 0 | 0.0%
Driver.DriverDepDynamic | 0 | 0 | 0 | 0.0%
Driver.DriverDepExternal | 0 | 0 | 0 | 0.0%
Driver.DriverDepMember | 0 | 0 | 0 | 0.0%
Driver.DriverDepNominal | 0 | 0 | 0 | 0.0%
Driver.DriverDepTopLevel | 0 | 0 | 0 | 0.0%
Driver.NumDriverJobsRun | 18,540 | 18,540 | 0 | 0.0%
Driver.NumDriverJobsSkipped | 0 | 0 | 0 | 0.0%
Driver.NumProcessFailures | 0 | 0 | 0 | 0.0%
Frontend.MaxMallocUsage | 732,496,213,920 | 731,593,165,432 | -903,048,488 | -0.12%
Frontend.NumInstructionsExecuted | 29,797,283,768,170 | 29,714,923,670,196 | -82,360,097,974 | -0.28%
Frontend.NumProcessFailures | 0 | 0 | 0 | 0.0%
IRModule.NumIRAliases | 115,571 | 115,571 | 0 | 0.0%
IRModule.NumIRBasicBlocks | 4,399,402 | 4,399,402 | 0 | 0.0%
IRModule.NumIRComdatSymbols | 0 | 0 | 0 | 0.0%
IRModule.NumIRFunctions | 2,083,526 | 2,083,526 | 0 | 0.0%
IRModule.NumIRGlobals | 2,107,811 | 2,107,811 | 0 | 0.0%
IRModule.NumIRIFuncs | 0 | 0 | 0 | 0.0%
IRModule.NumIRInsts | 54,867,102 | 54,867,102 | 0 | 0.0%
IRModule.NumIRNamedMetaData | 90,000 | 90,000 | 0 | 0.0%
IRModule.NumIRValueSymbols | 3,789,105 | 3,789,105 | 0 | 0.0%
LLVM.NumLLVMBytesOutput | 1,114,729,562 | 1,114,726,372 | -3,190 | -0.0%
Parse.NumFunctionsParsed | 168,448 | 168,448 | 0 | 0.0%
Parse.NumIterableDeclContextParsed | 1,163,266 | 1,163,266 | 0 | 0.0%
SILModule.NumSILGenDefaultWitnessTables | 0 | 0 | 0 | 0.0%
SILModule.NumSILGenFunctions | 1,071,861 | 1,071,861 | 0 | 0.0%
SILModule.NumSILGenGlobalVariables | 40,105 | 40,105 | 0 | 0.0%
SILModule.NumSILGenVtables | 11,745 | 11,745 | 0 | 0.0%
SILModule.NumSILGenWitnessTables | 45,902 | 45,902 | 0 | 0.0%
SILModule.NumSILOptDefaultWitnessTables | 0 | 0 | 0 | 0.0%
SILModule.NumSILOptFunctions | 1,499,596 | 1,499,596 | 0 | 0.0%
SILModule.NumSILOptGlobalVariables | 40,991 | 40,991 | 0 | 0.0%
SILModule.NumSILOptVtables | 19,254 | 19,254 | 0 | 0.0%
SILModule.NumSILOptWitnessTables | 99,776 | 99,776 | 0 | 0.0%
Sema.AccessLevelRequest | 2,769,441 | 2,769,100 | -341 | -0.01%
Sema.AttachedPropertyDelegateRequest | 1,399,808 | 1,399,808 | 0 | 0.0%
Sema.AttachedPropertyDelegateTypeRequest | 243,821 | 243,821 | 0 | 0.0%
Sema.CustomAttrNominalRequest | 0 | 0 | 0 | 0.0%
Sema.DefaultAndMaxAccessLevelRequest | 62,803 | 62,797 | -6 | -0.01%
Sema.DefaultTypeRequest | 375,224 | 375,224 | 0 | 0.0%
Sema.EnumRawTypeRequest | 20,062 | 20,062 | 0 | 0.0%
Sema.ExtendedNominalRequest | 4,091,288 | 4,085,259 | -6,029 | -0.15%
Sema.InheritedDeclsReferencedRequest | 4,782,389 | 4,755,738 | -26,651 | -0.56%
Sema.InheritedTypeRequest | 646,385 | 645,598 | -787 | -0.12%
Sema.IsDynamicRequest | 2,133,425 | 2,133,425 | 0 | 0.0%
Sema.IsFinalRequest | 3,915,191 | 3,903,832 | -11,359 | -0.29%
Sema.IsObjCRequest | 1,873,932 | 1,874,157 | 225 | 0.01%
Sema.MangleLocalTypeDeclRequest | 246 | 246 | 0 | 0.0%
Sema.NamedLazyMemberLoadFailureCount | 23,254 | 23,171 | -83 | -0.36%
Sema.NamedLazyMemberLoadSuccessCount | 20,098,329 | 20,093,381 | -4,948 | -0.02%
Sema.NominalTypeLookupDirectCount | 33,706,377 | 33,676,258 | -30,119 | -0.09%
Sema.NumConformancesDeserialized | 6,922,504 | 6,894,546 | -27,958 | -0.4%
Sema.NumConstraintScopes | 17,558,595 | 17,554,942 | -3,653 | -0.02%
Sema.NumConstraintsConsideredForEdgeContraction | 43,409,513 | 43,407,914 | -1,599 | -0.0%
Sema.NumDeclsDeserialized | 51,842,833 | 51,635,585 | -207,248 | -0.4%
Sema.NumDeclsFinalized | 1,937,023 | 1,937,023 | 0 | 0.0%
Sema.NumDeclsTypechecked | 1,003,406 | 1,003,406 | 0 | 0.0%
Sema.NumDeclsValidated | 2,350,190 | 2,350,196 | 6 | 0.0%
Sema.NumFunctionsTypechecked | 1,063,592 | 1,063,592 | 0 | 0.0%
Sema.NumGenericSignatureBuilders | 1,292,199 | 1,289,850 | -2,349 | -0.18%
Sema.NumLazyGenericEnvironments | 10,448,472 | 10,444,575 | -3,897 | -0.04%
Sema.NumLazyGenericEnvironmentsLoaded | 222,779 | 222,559 | -220 | -0.1%
Sema.NumLazyIterableDeclContexts | 7,251,362 | 7,241,413 | -9,949 | -0.14%
Sema.NumLeafScopes | 11,796,929 | 11,794,142 | -2,787 | -0.02%
Sema.NumTypesDeserialized | 17,530,876 | 17,497,872 | -33,004 | -0.19%
Sema.NumTypesValidated | 1,794,789 | 1,794,796 | 7 | 0.0%
Sema.NumUnloadedLazyIterableDeclContexts | 4,766,652 | 4,764,058 | -2,594 | -0.05%
Sema.OverriddenDeclsRequest | 8,139,958 | 8,097,603 | -42,355 | -0.52%
Sema.PropertyDelegateBackingPropertyInfoRequest | 241,304 | 241,304 | 0 | 0.0%
Sema.PropertyDelegateBackingPropertyTypeRequest | 243,821 | 243,821 | 0 | 0.0%
Sema.PropertyDelegateTypeInfoRequest | 0 | 0 | 0 | 0.0%
Sema.RequirementRequest | 71,878 | 71,879 | 1 | 0.0%
Sema.SelfBoundsFromWhereClauseRequest | 6,847,671 | 6,812,801 | -34,870 | -0.51%
Sema.SetterAccessLevelRequest | 157,229 | 157,229 | 0 | 0.0%
Sema.SuperclassDeclRequest | 86,202 | 85,902 | -300 | -0.35%
Sema.SuperclassTypeRequest | 37,133 | 37,133 | 0 | 0.0%
Sema.TypeDeclsFromWhereClauseRequest | 33,069 | 33,063 | -6 | -0.02%
Sema.USRGenerationRequest | 13,217,879 | 13,108,976 | -108,903 | -0.82%
Sema.UnderlyingTypeDeclsReferencedRequest | 185,226 | 184,808 | -418 | -0.23%
</details>

<a name="release-39e557b914ed98a1">

# Release

<a name="release-brief-39e557b914ed98a1">

## release brief
<details>
<summary>Regressed (0)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
</details>
<details>
<summary>Improved (0)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
</details>
<details>
<summary>Unchanged (delta < 1.0% or delta < 100.0ms) (3)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
Frontend.NumInstructionsExecuted | 28,885,541,413,014 | 28,896,294,066,250 | 10,752,653,236 | 0.04%
LLVM.NumLLVMBytesOutput | 944,140,438 | 944,143,858 | 3,420 | 0.0%
time.swift-driver.wall | 5063.7s | 5069.6s | 5.9s | 0.12%
</details>

<a name="release-detailed-39e557b914ed98a1">

## release detailed
<details>
<summary>Regressed (0)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
</details>
<details>
<summary>Improved (0)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
</details>
<details>
<summary>Unchanged (delta < 1.0% or delta < 100.0ms) (23)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
AST.NumImportedExternalDefinitions | 231,008 | 231,008 | 0 | 0.0%
AST.NumLoadedModules | 16,804 | 16,804 | 0 | 0.0%
AST.NumTotalClangImportedEntities | 792,912 | 792,912 | 0 | 0.0%
AST.NumUsedConformances | 261,370 | 261,370 | 0 | 0.0%
IRModule.NumIRBasicBlocks | 3,824,271 | 3,824,271 | 0 | 0.0%
IRModule.NumIRFunctions | 1,743,720 | 1,743,720 | 0 | 0.0%
IRModule.NumIRGlobals | 1,849,488 | 1,849,488 | 0 | 0.0%
IRModule.NumIRInsts | 34,233,063 | 34,233,047 | -16 | -0.0%
IRModule.NumIRValueSymbols | 3,360,268 | 3,360,268 | 0 | 0.0%
LLVM.NumLLVMBytesOutput | 944,140,438 | 944,143,858 | 3,420 | 0.0%
SILModule.NumSILGenFunctions | 743,967 | 743,967 | 0 | 0.0%
SILModule.NumSILOptFunctions | 997,831 | 997,831 | 0 | 0.0%
Sema.NumConformancesDeserialized | 2,329,983 | 2,329,983 | 0 | 0.0%
Sema.NumConstraintScopes | 15,703,927 | 15,703,927 | 0 | 0.0%
Sema.NumDeclsDeserialized | 6,282,743 | 6,282,743 | 0 | 0.0%
Sema.NumDeclsValidated | 1,208,430 | 1,208,430 | 0 | 0.0%
Sema.NumFunctionsTypechecked | 479,178 | 479,178 | 0 | 0.0%
Sema.NumGenericSignatureBuilders | 210,055 | 210,055 | 0 | 0.0%
Sema.NumLazyGenericEnvironments | 1,288,733 | 1,288,733 | 0 | 0.0%
Sema.NumLazyGenericEnvironmentsLoaded | 22,291 | 22,291 | 0 | 0.0%
Sema.NumLazyIterableDeclContexts | 797,418 | 797,418 | 0 | 0.0%
Sema.NumTypesDeserialized | 3,323,682 | 3,323,682 | 0 | 0.0%
Sema.NumTypesValidated | 734,764 | 734,764 | 0 | 0.0%
</details>

**Comment by gottesmm** [comment]
@slavapestov SGTM!

**Comment by compnerd** [comment]
Pretty sure that this broke the Linux builds.  It seems to have broke the windows build too:

```
S:\swift\lib\SILGen\SILGenFunction.cpp(559): error C2666: 'swift::SILBuilder::createApply': 2 overloads have similar conversions
S:\swift\include\swift/SIL/SILBuilder.h(462): note: could be 'swift::ApplyInst *swift::SILBuilder::createApply(swift::SILLocation,swift::SILValue,llvm::ArrayRef<swift::SILValue>,bool,const swift::GenericSpecializationInformation *)'
S:\swift\include\swift/SIL/SILBuilder.h(453): note: or       'swift::ApplyInst *swift::SILBuilder::createApply(swift::SILLocation,swift::SILValue,swift::SubstitutionMap,llvm::ArrayRef<swift::SILValue>,bool,const swift::GenericSpecializationInformation *)'
S:\swift\lib\SILGen\SILGenFunction.cpp(559): note: while trying to match the argument list '(swift::ClassDecl *, swift::FunctionRefInst *, initializer list, swift::SILValue [4], bool)'
```

**Comment by compnerd** [comment]
#24276 should fix it

**Comment by slavapestov** [comment]
Not sure how it could have broken Linux builds since it passed CI. But thanks for the fix!

**Pull request merged**